File: resources/ios/Sources/TabBarFunctions.swift
import Foundation
import UIKit

// MARK: - Tab Bar Manager (Singleton)

class TabBarManager {
    static let shared = TabBarManager()

    private var tabBar: UITabBar?
    private var tabBarBottomConstraint: NSLayoutConstraint?
    private var webViewBottomConstraint: NSLayoutConstraint?
    private var tabConfigs: [[String: Any]] = []
    private var styleConfig: [String: Any] = [:]
    private var isVisible: Bool = true

    private init() {}

    // MARK: - Configuration

    func configure(parameters: [String: Any]) {
        guard let tabs = parameters["tabs"] as? [[String: Any]] else { return }

        self.tabConfigs = tabs
        self.styleConfig = parameters["style"] as? [String: Any] ?? [:]

        DispatchQueue.main.async {
            self.setupTabBar()
            self.applyStyle()
            self.applyBadges()

            // Set active tab
            if let activeTabId = parameters["active_tab"] as? String {
                self.selectTab(byId: activeTabId)
            } else {
                // Default to first tab
                self.tabBar?.selectedItem = self.tabBar?.items?.first
            }
        }
    }

    func update(parameters: [String: Any]) {
        guard let tabs = parameters["tabs"] as? [[String: Any]] else { return }

        self.tabConfigs = tabs

        if let style = parameters["style"] as? [String: Any] {
            self.styleConfig = style
        }

        DispatchQueue.main.async {
            self.rebuildTabItems()
            self.applyStyle()
            self.applyBadges()

            if let activeTabId = parameters["active_tab"] as? String {
                self.selectTab(byId: activeTabId)
            }
        }
    }

    func setActive(tabId: String) {
        DispatchQueue.main.async {
            self.selectTab(byId: tabId)
        }
    }

    func setBadge(tabId: String, count: Int?) {
        DispatchQueue.main.async {
            guard let items = self.tabBar?.items else { return }
            guard let index = self.tabConfigs.firstIndex(where: { $0["id"] as? String == tabId }) else { return }
            guard index < items.count else { return }

            if let count = count, count > 0 {
                items[index].badgeValue = "\(count)"
            } else {
                items[index].badgeValue = nil
            }
        }
    }

    func show() {
        guard !isVisible else { return }
        isVisible = true

        DispatchQueue.main.async {
            UIView.animate(withDuration: 0.3) {
                self.tabBar?.isHidden = false
                self.tabBarBottomConstraint?.constant = 0
                self.tabBar?.superview?.layoutIfNeeded()
            }
        }
    }

    func hide() {
        guard isVisible else { return }
        isVisible = false

        DispatchQueue.main.async {
            UIView.animate(withDuration: 0.3) {
                self.tabBarBottomConstraint?.constant = 100
                self.tabBar?.superview?.layoutIfNeeded()
            } completion: { _ in
                self.tabBar?.isHidden = true
            }
        }
    }

    // MARK: - Private Setup

    private func setupTabBar() {
        // Remove existing tab bar if reconfiguring
        tabBar?.removeFromSuperview()

        guard let window = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .flatMap({ $0.windows })
            .first(where: { $0.isKeyWindow }),
              let rootView = window.rootViewController?.view else { return }

        let bar = UITabBar()
        bar.delegate = TabBarDelegateHandler.shared
        bar.translatesAutoresizingMaskIntoConstraints = false

        rootView.addSubview(bar)

        let bottomConstraint = bar.bottomAnchor.constraint(equalTo: rootView.safeAreaLayoutGuide.bottomAnchor)

        NSLayoutConstraint.activate([
            bar.leadingAnchor.constraint(equalTo: rootView.leadingAnchor),
            bar.trailingAnchor.constraint(equalTo: rootView.trailingAnchor),
            bottomConstraint,
        ])

        self.tabBar = bar
        self.tabBarBottomConstraint = bottomConstraint

        rebuildTabItems()
    }

    private func rebuildTabItems() {
        var items: [UITabBarItem] = []

        for (index, config) in tabConfigs.enumerated() {
            let label = config["label"] as? String ?? ""
            let iconName = config["icon"] as? String ?? ""
            let iconType = config["icon_type"] as? String ?? "system"
            let visible = config["visible"] as? Bool ?? true

            guard visible else { continue }

            let image: UIImage?
            if iconType == "custom" {
                image = UIImage(named: iconName)
            } else {
                image = UIImage(systemName: iconName)
            }

            // Active icon
            var selectedImage: UIImage? = nil
            if let activeIconName = config["active_icon"] as? String {
                let activeIconType = config["active_icon_type"] as? String ?? "system"
                if activeIconType == "custom" {
                    selectedImage = UIImage(named: activeIconName)
                } else {
                    selectedImage = UIImage(systemName: activeIconName)
                }
            }

            let item = UITabBarItem(title: label, image: image, selectedImage: selectedImage)
            item.tag = index

            items.append(item)
        }

        tabBar?.setItems(items, animated: false)
    }

    private func applyStyle() {
        guard let bar = tabBar else { return }

        let isDarkMode = bar.traitCollection.userInterfaceStyle == .dark
        let darkStyle = styleConfig["dark"] as? [String: Any] ?? [:]

        // Resolve color based on current appearance
        func resolveColor(_ lightKey: String) -> UIColor? {
            if isDarkMode, let darkHex = darkStyle[lightKey] as? String {
                return UIColor(hex: darkHex)
            }
            if let lightHex = styleConfig[lightKey] as? String {
                return UIColor(hex: lightHex)
            }
            return nil
        }

        let appearance = UITabBarAppearance()

        if let translucent = styleConfig["translucent"] as? Bool, !translucent {
            appearance.configureWithOpaqueBackground()
        } else {
            appearance.configureWithDefaultBackground()
        }

        if let bgColor = resolveColor("background_color") {
            appearance.backgroundColor = bgColor
        }

        if let activeColor = resolveColor("active_color") {
            bar.tintColor = activeColor
        }

        if let inactiveColor = resolveColor("inactive_color") {
            appearance.stackedLayoutAppearance.normal.iconColor = inactiveColor
            appearance.stackedLayoutAppearance.normal.titleTextAttributes = [.foregroundColor: inactiveColor]
        }

        if let borderColor = resolveColor("border_color") {
            appearance.shadowColor = borderColor
        }

        // Badge styling
        if let badgeColor = resolveColor("badge_color") {
            appearance.stackedLayoutAppearance.normal.badgeBackgroundColor = badgeColor
            appearance.stackedLayoutAppearance.selected.badgeBackgroundColor = badgeColor
        }

        if let badgeTextColor = resolveColor("badge_text_color") {
            appearance.stackedLayoutAppearance.normal.badgeTextAttributes = [.foregroundColor: badgeTextColor]
            appearance.stackedLayoutAppearance.selected.badgeTextAttributes = [.foregroundColor: badgeTextColor]
        }

        bar.standardAppearance = appearance
        bar.scrollEdgeAppearance = appearance
    }

    private func applyBadges() {
        guard let items = tabBar?.items else { return }

        for (index, config) in tabConfigs.enumerated() {
            guard index < items.count else { break }

            if let badge = config["badge"] as? Int, badge > 0 {
                items[index].badgeValue = "\(badge)"

                if let badgeColor = config["badge_color"] as? String {
                    items[index].badgeColor = UIColor(hex: badgeColor)
                }
            }
        }
    }

    private func selectTab(byId tabId: String) {
        guard let items = tabBar?.items else { return }
        guard let index = tabConfigs.firstIndex(where: { $0["id"] as? String == tabId }) else { return }
        guard index < items.count else { return }

        tabBar?.selectedItem = items[index]
    }

    // MARK: - Tab Tap Handling (called by delegate)

    func handleTabTap(at index: Int) {
        guard index < tabConfigs.count else { return }

        let config = tabConfigs[index]
        let id = config["id"] as? String ?? ""
        let type = config["type"] as? String ?? "url"
        let url = config["url"] as? String
        let action = config["action"] as? String

        // Dispatch TabSelected event for all taps
        let selectedPayload: [String: Any] = [
            "id": id,
            "url": url ?? "",
            "index": index,
        ]

        LaravelBridge.shared.send?(
            "Krakero\\TabBar\\Events\\TabSelected",
            selectedPayload
        )

        if type == "url", let url = url {
            // Navigate the webview
            navigateWebView(to: url)
        } else if type == "action", let action = action {
            // Dispatch action event
            let actionPayload: [String: Any] = [
                "id": id,
                "action": action,
            ]

            LaravelBridge.shared.send?(
                "Krakero\\TabBar\\Events\\TabActionTriggered",
                actionPayload
            )
        }
    }

    private func navigateWebView(to url: String) {
        // Inject JavaScript to navigate the webview
        guard let window = UIApplication.shared.connectedScenes
            .compactMap({ $0 as? UIWindowScene })
            .flatMap({ $0.windows })
            .first(where: { $0.isKeyWindow }),
              let rootVC = window.rootViewController else { return }

        // Find the WKWebView in the view hierarchy
        if let webView = findWebView(in: rootVC.view) {
            let js = "window.location.href = '\(url)';"
            webView.evaluateJavaScript(js, completionHandler: nil)
        }
    }

    private func findWebView(in view: UIView) -> WKWebView? {
        if let webView = view as? WKWebView {
            return webView
        }
        for subview in view.subviews {
            if let found = findWebView(in: subview) {
                return found
            }
        }
        return nil
    }
}

// MARK: - Tab Bar Delegate

class TabBarDelegateHandler: NSObject, UITabBarDelegate {
    static let shared = TabBarDelegateHandler()

    func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem) {
        TabBarManager.shared.handleTabTap(at: item.tag)
    }
}

// MARK: - UIColor Hex Extension

extension UIColor {
    convenience init?(hex: String) {
        var hexSanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        hexSanitized = hexSanitized.replacingOccurrences(of: "#", with: "")

        var rgb: UInt64 = 0
        guard Scanner(string: hexSanitized).scanHexInt64(&rgb) else { return nil }

        let length = hexSanitized.count
        if length == 6 {
            self.init(
                red: CGFloat((rgb & 0xFF0000) >> 16) / 255.0,
                green: CGFloat((rgb & 0x00FF00) >> 8) / 255.0,
                blue: CGFloat(rgb & 0x0000FF) / 255.0,
                alpha: 1.0
            )
        } else if length == 8 {
            self.init(
                red: CGFloat((rgb & 0xFF000000) >> 24) / 255.0,
                green: CGFloat((rgb & 0x00FF0000) >> 16) / 255.0,
                blue: CGFloat((rgb & 0x0000FF00) >> 8) / 255.0,
                alpha: CGFloat(rgb & 0x000000FF) / 255.0
            )
        } else {
            return nil
        }
    }
}

// MARK: - WKWebView Import

import WebKit

// MARK: - Bridge Functions

enum TabBarFunctions {

    class Configure: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            TabBarManager.shared.configure(parameters: parameters)
            return BridgeResponse.success(data: ["configured": true])
        }
    }

    class Update: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            TabBarManager.shared.update(parameters: parameters)
            return BridgeResponse.success(data: ["updated": true])
        }
    }

    class SetActive: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let id = parameters["id"] as? String else {
                return BridgeResponse.error(message: "Missing required parameter: id")
            }
            TabBarManager.shared.setActive(tabId: id)
            return BridgeResponse.success(data: ["active": id])
        }
    }

    class SetBadge: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard let id = parameters["id"] as? String else {
                return BridgeResponse.error(message: "Missing required parameter: id")
            }
            let count = parameters["count"] as? Int
            TabBarManager.shared.setBadge(tabId: id, count: count)
            return BridgeResponse.success(data: ["id": id, "badge": count ?? 0])
        }
    }

    class Show: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            TabBarManager.shared.show()
            return BridgeResponse.success(data: ["visible": true])
        }
    }

    class Hide: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            TabBarManager.shared.hide()
            return BridgeResponse.success(data: ["visible": false])
        }
    }
}

File: resources/android/src/TabBarFunctions.kt
package com.krakero.plugins.tabbar

import android.app.Activity
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.view.Gravity
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.FrameLayout
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.badge.BadgeDrawable
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import org.json.JSONObject

// ============================================================
// Tab Bar Manager (Singleton)
// ============================================================

object TabBarManager {

    private var bottomNav: BottomNavigationView? = null
    private var tabConfigs: List<Map<String, Any>> = emptyList()
    private var styleConfig: Map<String, Any> = emptyMap()
    private var isVisible: Boolean = true
    private var activity: Activity? = null

    // SF Symbol name -> Material Icon mapping
    private val iconMap = mapOf(
        "house" to android.R.drawable.ic_menu_recent_history,
        "house.fill" to android.R.drawable.ic_menu_recent_history,
        "magnifyingglass" to android.R.drawable.ic_menu_search,
        "plus.circle" to android.R.drawable.ic_menu_add,
        "plus.circle.fill" to android.R.drawable.ic_menu_add,
        "bell" to android.R.drawable.ic_popup_reminder,
        "bell.fill" to android.R.drawable.ic_popup_reminder,
        "person.circle" to android.R.drawable.ic_menu_myplaces,
        "person.circle.fill" to android.R.drawable.ic_menu_myplaces,
        "gear" to android.R.drawable.ic_menu_preferences,
        "gearshape" to android.R.drawable.ic_menu_preferences,
        "star" to android.R.drawable.btn_star_big_off,
        "star.fill" to android.R.drawable.btn_star_big_on,
        "heart" to android.R.drawable.ic_menu_close_clear_cancel,
        "heart.fill" to android.R.drawable.ic_menu_close_clear_cancel,
        "envelope" to android.R.drawable.ic_dialog_email,
        "camera" to android.R.drawable.ic_menu_camera,
        "camera.fill" to android.R.drawable.ic_menu_camera,
        "map" to android.R.drawable.ic_menu_mapmode,
        "info.circle" to android.R.drawable.ic_menu_info_details,
        "trash" to android.R.drawable.ic_menu_delete,
        "pencil" to android.R.drawable.ic_menu_edit,
        "square.and.arrow.up" to android.R.drawable.ic_menu_share,
    )

    // ----------------------------------------------------------
    // Public API
    // ----------------------------------------------------------

    fun configure(activity: Activity, parameters: Map<String, Any>) {
        this.activity = activity

        @Suppress("UNCHECKED_CAST")
        val tabs = parameters["tabs"] as? List<Map<String, Any>> ?: return
        this.tabConfigs = tabs
        @Suppress("UNCHECKED_CAST")
        this.styleConfig = parameters["style"] as? Map<String, Any> ?: emptyMap()

        Handler(Looper.getMainLooper()).post {
            setupBottomNav(activity)
            applyStyle()
            applyBadges()

            val activeTabId = parameters["active_tab"] as? String
            if (activeTabId != null) {
                selectTab(activeTabId)
            }
        }
    }

    fun update(activity: Activity, parameters: Map<String, Any>) {
        this.activity = activity

        @Suppress("UNCHECKED_CAST")
        val tabs = parameters["tabs"] as? List<Map<String, Any>> ?: return
        this.tabConfigs = tabs

        @Suppress("UNCHECKED_CAST")
        val style = parameters["style"] as? Map<String, Any>
        if (style != null) {
            this.styleConfig = style
        }

        Handler(Looper.getMainLooper()).post {
            rebuildMenu()
            applyStyle()
            applyBadges()

            val activeTabId = parameters["active_tab"] as? String
            if (activeTabId != null) {
                selectTab(activeTabId)
            }
        }
    }

    fun setActive(tabId: String) {
        Handler(Looper.getMainLooper()).post {
            selectTab(tabId)
        }
    }

    fun setBadge(tabId: String, count: Int?) {
        Handler(Looper.getMainLooper()).post {
            val nav = bottomNav ?: return@post
            val index = tabConfigs.indexOfFirst { it["id"] == tabId }
            if (index < 0) return@post

            val menuItem = nav.menu.getItem(index) ?: return@post

            if (count != null && count > 0) {
                val badge = nav.getOrCreateBadge(menuItem.itemId)
                badge.number = count
                badge.isVisible = true
            } else {
                nav.removeBadge(menuItem.itemId)
            }
        }
    }

    fun show() {
        if (isVisible) return
        isVisible = true

        Handler(Looper.getMainLooper()).post {
            bottomNav?.visibility = View.VISIBLE
        }
    }

    fun hide() {
        if (!isVisible) return
        isVisible = false

        Handler(Looper.getMainLooper()).post {
            bottomNav?.visibility = View.GONE
        }
    }

    // ----------------------------------------------------------
    // Private Setup
    // ----------------------------------------------------------

    private fun setupBottomNav(activity: Activity) {
        // Remove existing nav if reconfiguring
        bottomNav?.let {
            (it.parent as? ViewGroup)?.removeView(it)
        }

        val nav = BottomNavigationView(activity)
        nav.id = View.generateViewId()
        nav.labelVisibilityMode = BottomNavigationView.LABEL_VISIBILITY_LABELED

        // Add to root view
        val rootView = activity.findViewById<ViewGroup>(android.R.id.content)
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.MATCH_PARENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        )
        params.gravity = Gravity.BOTTOM
        rootView.addView(nav, params)

        // Handle tab selection
        nav.setOnItemSelectedListener { menuItem ->
            val index = indexForMenuItemId(menuItem.itemId)
            if (index >= 0) {
                handleTabTap(activity, index)
            }
            true
        }

        this.bottomNav = nav
        rebuildMenu()
    }

    private fun rebuildMenu() {
        val nav = bottomNav ?: return
        nav.menu.clear()

        for ((index, config) in tabConfigs.withIndex()) {
            val label = config["label"] as? String ?: ""
            val iconName = config["icon"] as? String ?: ""
            val iconType = config["icon_type"] as? String ?: "system"
            val visible = config["visible"] as? Boolean ?: true

            if (!visible) continue

            val iconRes = if (iconType == "custom") {
                // Look up custom drawable by name
                activity?.resources?.getIdentifier(iconName, "drawable", activity?.packageName) ?: 0
            } else {
                // Map SF Symbol name to Android resource
                iconMap[iconName] ?: android.R.drawable.ic_menu_help
            }

            val menuItem = nav.menu.add(Menu.NONE, index, index, label)
            menuItem.setIcon(iconRes)
        }
    }

    private fun applyStyle() {
        val nav = bottomNav ?: return
        val isDarkMode = (nav.resources.configuration.uiMode and
                android.content.res.Configuration.UI_MODE_NIGHT_MASK) ==
                android.content.res.Configuration.UI_MODE_NIGHT_YES

        @Suppress("UNCHECKED_CAST")
        val darkStyle = styleConfig["dark"] as? Map<String, Any> ?: emptyMap()

        fun resolveColor(key: String): Int? {
            if (isDarkMode) {
                val darkHex = darkStyle[key] as? String
                if (darkHex != null) return parseColor(darkHex)
            }
            val lightHex = styleConfig[key] as? String
            if (lightHex != null) return parseColor(lightHex)
            return null
        }

        // Background
        resolveColor("background_color")?.let { nav.setBackgroundColor(it) }

        // Active / inactive icon and text colors
        val activeColor = resolveColor("active_color")
        val inactiveColor = resolveColor("inactive_color")

        if (activeColor != null || inactiveColor != null) {
            val states = arrayOf(
                intArrayOf(android.R.attr.state_checked),
                intArrayOf(-android.R.attr.state_checked)
            )
            val colors = intArrayOf(
                activeColor ?: Color.parseColor("#007AFF"),
                inactiveColor ?: Color.parseColor("#8E8E93")
            )
            val colorStateList = ColorStateList(states, colors)
            nav.itemIconTintList = colorStateList
            nav.itemTextColor = colorStateList
        }

        // Elevation
        val elevation = styleConfig["elevation"]
        if (elevation is Number) {
            nav.elevation = elevation.toFloat() * nav.resources.displayMetrics.density
        }

        // Badge colors
        resolveColor("badge_color")?.let { badgeColor ->
            for (i in 0 until nav.menu.size()) {
                val badge = nav.getBadge(nav.menu.getItem(i).itemId) ?: continue
                badge.backgroundColor = badgeColor
            }
        }

        resolveColor("badge_text_color")?.let { textColor ->
            for (i in 0 until nav.menu.size()) {
                val badge = nav.getBadge(nav.menu.getItem(i).itemId) ?: continue
                badge.badgeTextColor = textColor
            }
        }
    }

    private fun applyBadges() {
        val nav = bottomNav ?: return

        for ((index, config) in tabConfigs.withIndex()) {
            if (index >= nav.menu.size()) break

            val badge = config["badge"]
            val menuItem = nav.menu.getItem(index)

            if (badge is Number && badge.toInt() > 0) {
                val badgeDrawable = nav.getOrCreateBadge(menuItem.itemId)
                badgeDrawable.number = badge.toInt()
                badgeDrawable.isVisible = true

                val badgeColor = config["badge_color"] as? String
                if (badgeColor != null) {
                    parseColor(badgeColor)?.let { badgeDrawable.backgroundColor = it }
                }
            }
        }
    }

    private fun selectTab(tabId: String) {
        val nav = bottomNav ?: return
        val index = tabConfigs.indexOfFirst { it["id"] == tabId }
        if (index < 0 || index >= nav.menu.size()) return

        nav.selectedItemId = nav.menu.getItem(index).itemId
    }

    private fun indexForMenuItemId(itemId: Int): Int {
        return itemId // We use index as the menu item ID
    }

    // ----------------------------------------------------------
    // Tab Tap Handling
    // ----------------------------------------------------------

    private fun handleTabTap(activity: Activity, index: Int) {
        if (index >= tabConfigs.size) return

        val config = tabConfigs[index]
        val id = config["id"] as? String ?: ""
        val type = config["type"] as? String ?: "url"
        val url = config["url"] as? String
        val action = config["action"] as? String

        // Dispatch TabSelected event for all taps
        val selectedPayload = JSONObject().apply {
            put("id", id)
            put("url", url ?: "")
            put("index", index)
        }

        Handler(Looper.getMainLooper()).post {
            NativeActionCoordinator.dispatchEvent(
                activity,
                "Krakero\\TabBar\\Events\\TabSelected",
                selectedPayload.toString()
            )
        }

        if (type == "url" && url != null) {
            // Navigate the webview
            navigateWebView(activity, url)
        } else if (type == "action" && action != null) {
            // Dispatch action event
            val actionPayload = JSONObject().apply {
                put("id", id)
                put("action", action)
            }

            Handler(Looper.getMainLooper()).post {
                NativeActionCoordinator.dispatchEvent(
                    activity,
                    "Krakero\\TabBar\\Events\\TabActionTriggered",
                    actionPayload.toString()
                )
            }
        }
    }

    private fun navigateWebView(activity: Activity, url: String) {
        Handler(Looper.getMainLooper()).post {
            val rootView = activity.findViewById<ViewGroup>(android.R.id.content)
            val webView = findWebView(rootView)
            webView?.evaluateJavascript("window.location.href = '$url';", null)
        }
    }

    private fun findWebView(view: View): WebView? {
        if (view is WebView) return view
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                val found = findWebView(view.getChildAt(i))
                if (found != null) return found
            }
        }
        return null
    }

    // ----------------------------------------------------------
    // Utility
    // ----------------------------------------------------------

    private fun parseColor(hex: String): Int? {
        return try {
            Color.parseColor(if (hex.startsWith("#")) hex else "#$hex")
        } catch (e: Exception) {
            null
        }
    }
}

// ============================================================
// Bridge Functions
// ============================================================

object TabBarFunctions {

    class Configure : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val activity = NativeActionCoordinator.getCurrentActivity()
                ?: return BridgeResponse.error("No active activity found")

            TabBarManager.configure(activity, parameters)
            return BridgeResponse.success(mapOf("configured" to true))
        }
    }

    class Update : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val activity = NativeActionCoordinator.getCurrentActivity()
                ?: return BridgeResponse.error("No active activity found")

            TabBarManager.update(activity, parameters)
            return BridgeResponse.success(mapOf("updated" to true))
        }
    }

    class SetActive : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val id = parameters["id"] as? String
                ?: return BridgeResponse.error("Missing required parameter: id")

            TabBarManager.setActive(id)
            return BridgeResponse.success(mapOf("active" to id))
        }
    }

    class SetBadge : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            val id = parameters["id"] as? String
                ?: return BridgeResponse.error("Missing required parameter: id")

            val count = (parameters["count"] as? Number)?.toInt()
            TabBarManager.setBadge(id, count)
            return BridgeResponse.success(mapOf("id" to id, "badge" to (count ?: 0)))
        }
    }

    class Show : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            TabBarManager.show()
            return BridgeResponse.success(mapOf("visible" to true))
        }
    }

    class Hide : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            TabBarManager.hide()
            return BridgeResponse.success(mapOf("visible" to false))
        }
    }
}
